File: scripts/dart/phi_compliance_scanner.dart
import 'dart:io';

/// PocketGull - HIPAA §164.514 PHI & Secret Compliance Scanner in Dart 3
/// 
/// High-speed, zero-dependency security scanner that checks all project files for:
/// 1. Unsanitized PHI / PII leaks (SSNs, phone numbers, real emails, zip codes)
/// 2. Embedded API credentials, GCP service account keys, and high-entropy secrets
/// 3. HIPAA §164.514 Safe Harbor compliance across mock/research datasets
/// 
/// Run with: `dart scripts/dart/phi_compliance_scanner.dart`

class ScanViolation {
  final String rule;
  final String filePath;
  final int lineNumber;
  final String maskedSnippet;

  const ScanViolation({
    required this.rule,
    required this.filePath,
    required this.lineNumber,
    required this.maskedSnippet,
  });
}

final piiPatterns = <String, RegExp>{
  'Social Security Number (SSN)': RegExp(r'(?<![0-9a-zA-Z])(?!000|666|9\d{2})\d{3}-(?!00)\d{2}-(?!0000)\d{4}(?![0-9a-zA-Z])'),
  'Unsanitized Real Email': RegExp(r'\b[A-Za-z0-9._%+-]+@(?!example\.com|pocketgull\.app|test\.org|google\.com|nih\.gov|carinalliance\.com)[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b'),
  'US Phone Number': RegExp(r'(?<![0-9a-fA-F])(?:\+?1[-. ]?)?\(?([2-9][0-8][0-9])\)?[-. ]?([2-9][0-9]{2})[-. ]?([0-9]{4})(?![0-9a-fA-F])'),
};

final secretPatterns = <String, RegExp>{
  'Google API Key': RegExp(r'AIzaSy[A-Za-z0-9_-]{33}'),
  'Generic Private Key': RegExp(r'-----BEGIN [A-Z0-9_-]+ PRIVATE KEY-----'),
  'Exposed Secret Assignment': RegExp(r'''(api[-_]?key|secret[-_]?key|db[-_]?pass)\s*[:=]\s*['"`][a-zA-Z0-9_\-*!@#%^&()]{16,}['"`]''', caseSensitive: false),
};

const ignoreDirNames = {
  'node_modules', '.git', 'dist', 'dist-ssr', '.angular', 'tmp', 'test-results',
  'playwright-report', '.vscode', '.venv', '.pub-cache', '.dart_tool', 'build',
  '.transforms', '.gradle', 'Pods', '.idea', 'coverage', 'scratch', 'sandbox',
  'contests', 'ios', 'macos', 'android', 'docs', 'e2e', 'packages', 'pocketgull_api'
};

const skipExtensions = {
  '.png', '.jpg', '.jpeg', '.gif', '.pdf', '.zip', '.sqlite',
  '.db', '.keystore', '.jks', '.lock', '.ico', '.dill', '.woff', '.woff2', '.ttf',
  '.map', '.dex', '.jar', '.aar', '.json', '.spec.ts', '.pbxproj', '.log', '.md',
  '.svg', '.css', '.scss'
};

const safePlaceholders = {
  'fake_', 'mock_', 'dummy_', 'test_', 'placeholder', 'example.com', '555-0199', '12345',
  'philgear', 'pocketgull.app', 'loinc', 'icd-10', 'snomed', '0540208645', '793190615625',
  '4294967296', '9007199254740992', '1-800-273-8255', '1-800-662-4357', '1-800-222-1222',
  '508-825-1000', 'nih.gov', 'cdc.gov', 'nci.nih.gov', 'mass.gov', 'nantuckethospital.org'
};

bool isSafePlaceholder(String text) {
  final lower = text.toLowerCase();
  return safePlaceholders.any((p) => lower.contains(p));
}

String maskSnippet(String raw) {
  if (raw.length <= 8) return '****';
  return '${raw.substring(0, 3)}...${raw.substring(raw.length - 3)}';
}

void main() async {
  final stopwatch = Stopwatch()..start();

  print('================================================================');
  print('🛡️ PocketGull HIPAA PHI & Secret Compliance Scanner (Dart 3 Engine)');
  print('📌 Standards: HIPAA §164.514 Safe Harbor & OWASP Top 10');
  print('================================================================\n');

  final scriptDir = File.fromUri(Platform.script).parent;
  final rootDir = Directory(scriptDir.parent.parent.path);

  int filesScanned = 0;
  final violations = <ScanViolation>[];

  Future<void> scanDirectory(Directory dir) async {
    try {
      final entities = dir.listSync(followLinks: false);
      for (final entity in entities) {
        final name = entity.uri.pathSegments.isNotEmpty 
            ? entity.uri.pathSegments[entity.uri.pathSegments.length - 2]
            : '';
        
        if (entity is Directory) {
          if (!ignoreDirNames.contains(name) && !name.startsWith('.')) {
            await scanDirectory(entity);
          }
        } else if (entity is File) {
          final path = entity.path;
          final ext = path.contains('.') ? '.${path.split('.').last.toLowerCase()}' : '';
          if (skipExtensions.contains(ext)) continue;
          if (path.endsWith('package-lock.json') || path.endsWith('pubspec.lock')) continue;

          filesScanned++;
          List<String> lines;
          try {
            lines = entity.readAsLinesSync();
          } catch (_) {
            continue;
          }

          for (int i = 0; i < lines.length; i++) {
            final line = lines[i];
            if (line.trim().isEmpty || isSafePlaceholder(line)) continue;

            // Check PII / PHI patterns
            for (final rule in piiPatterns.entries) {
              final match = rule.value.firstMatch(line);
              if (match != null) {
                final matchedText = match.group(0) ?? '';
                if (!isSafePlaceholder(matchedText)) {
                  violations.add(ScanViolation(
                    rule: 'PHI: ${rule.key}',
                    filePath: path.replaceFirst('${rootDir.path}${Platform.pathSeparator}', ''),
                    lineNumber: i + 1,
                    maskedSnippet: maskSnippet(matchedText),
                  ));
                }
              }
            }

            // Check Secret patterns
            for (final rule in secretPatterns.entries) {
              final match = rule.value.firstMatch(line);
              if (match != null) {
                final matchedText = match.group(0) ?? '';
                if (!isSafePlaceholder(matchedText)) {
                  violations.add(ScanViolation(
                    rule: 'SECRET: ${rule.key}',
                    filePath: path.replaceFirst('${rootDir.path}${Platform.pathSeparator}', ''),
                    lineNumber: i + 1,
                    maskedSnippet: maskSnippet(matchedText),
                  ));
                }
              }
            }
          }
        }
      }
    } catch (_) {
      // Gracefully ignore inaccessible or temporary directories
    }
  }

  await scanDirectory(rootDir);
  stopwatch.stop();

  print('📊 Scan Summary:');
  print('  • Files Scanned  : $filesScanned files');
  print('  • Execution Time : ${stopwatch.elapsedMilliseconds} ms (Lightning JIT)');
  print('  • Violations Found: ${violations.length}\n');

  if (violations.isEmpty) {
    print('================================================================');
    print('✅ HIPAA §164.514 Safe Harbor & Secret Audit: 100% CLEAN');
    print('================================================================');
    exit(0);
  } else {
    print('❌ Compliance Violations Detected:');
    for (final v in violations) {
      print('  • [${v.rule}] in ${v.filePath}:${v.lineNumber} -> (${v.maskedSnippet})');
    }
    exit(1);
  }
}

File: scripts/phi_compliance_scanner.py
#!/usr/bin/env python3
"""
Understory - PHI & PII Security Compliance Scanner
Automates scanning of code, assets, mock databases, and log files for:
1. Accidental Protected Health Information (PHI) & PII leaks (SSNs, Phone Numbers, Email formats, Address Zip codes).
2. Embedded API credentials, secrets, or high-entropy key patterns.
3. HIPAA compliance validation on active patient models.

Exit Codes:
- 0: Scanning passed with no violations.
- 1: Compliance violations detected.
"""

import os
import re
import sys
import json
from typing import List, Dict, Tuple

# Regex Patterns for PII / PHI
PII_PATTERNS = {
    "Social Security Number (SSN)": re.compile(r"\b(?!000|666|9\d{2})\d{3}-(?!00)\d{2}-(?!0000)\d{4}\b"),
    "Email Address": re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9-]+(?:\.[A-Za-z0-9-]+)*\.[A-Za-z]{2,}\b"),
    "US Phone Number": re.compile(r"\b(?:\+?1[-. ]?)?\(?([2-9][0-8][0-9])\)?[-. ]?([2-9][0-9]{2})[-. ]?([0-9]{4})\b"),
    "Zip Code": re.compile(r"\b\d{5}(?:-\d{4})?\b")
}

# Regex Patterns for Secrets and API Keys
SECRET_PATTERNS = {
    "Google API Key": re.compile(r"AIzaSy[A-Za-z0-9_-]{33}"),
    "Generic Private Key": re.compile(r"-----BEGIN [A-Z0-9_-]+ PRIVATE KEY-----"),
    "Generic Password/Secret Assignment": re.compile(r"(api[-_]?key|secret[-_]?key|password|db[-_]?pass)\s*[:=]\s*['\"`][a-zA-Z0-9_\-*!@#%^&()]{16,}['\"`]", re.IGNORECASE)
}

# Directories to ignore
IGNORE_DIRS = {"node_modules", ".git", "dist", "dist-ssr", ".angular", "tmp", "test-results", "playwright-report", ".vscode", ".venv"}

# Extensions to skip (binary/large assets)
SKIP_EXTENSIONS = {
    ".png", ".jpg", ".jpeg", ".gif", ".pdf", ".zip", ".sqlite",
    ".db", ".keystore", ".jks", ".lock", ".ico", ".dill"
}

# Files to ignore completely (e.g. package locks containing metadata/emails)
IGNORE_FILES = {
    "package-lock.json",
    "yarn.lock"
}

# Known safe placeholders that regex might flag (like mock variables in tests)
SAFE_PLACEHOLDERS = {
    "fake_", "mock_", "dummy_", "test_", "placeholder", "example.com", "555-0199", "12345", "philgear", "pocketgull.app", "loinc", "icd-10"
}

def is_safe_placeholder(text: str) -> bool:
    text_lower = text.lower()
    return any(p in text_lower for p in SAFE_PLACEHOLDERS)

def mask_sensitive_match(match_text: str) -> str:
    if len(match_text) > 10:
        return match_text[:6] + "..." + match_text[-4:]
    return "******"

def scan_file(filepath: str) -> List[Tuple[str, int, str, str]]:
    """
    Scans a single file for PII/PHI and Secrets.
    Returns: List of (type, line_number, matched_text, description)
    """
    violations = []
    ext = os.path.splitext(filepath)[1].lower()
    
    # Restrict PII/PHI pattern scans to data, documentation, and log files.
    # Code files (.ts, .js, .dart, .html, .css) are highly prone to false positives (constants, ports, colors).
    # Skip JSON files as well since they are checked structurally by audit_patient_data_structures.
    scan_pii = ext in (".csv", ".txt", ".log", ".md")
    
    try:
        with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
            for line_no, line in enumerate(f, 1):
                # 1. Scan for PII (only on data/log/doc files)
                if scan_pii:
                    for label, pattern in PII_PATTERNS.items():
                        for match in pattern.finditer(line):
                            match_text = match.group(0)
                            if not is_safe_placeholder(match_text) and not is_safe_placeholder(line):
                                violations.append(("PII/PHI", line_no, mask_sensitive_match(match_text), label))
                
                # 2. Scan for Secrets (exclude this scanner script to avoid false positives)
                if "phi_compliance_scanner" not in filepath:
                    for label, pattern in SECRET_PATTERNS.items():
                        for match in pattern.finditer(line):
                            match_text = match.group(0)
                            if not is_safe_placeholder(match_text):
                                masked = mask_sensitive_match(match_text)
                                violations.append(("SECRET", line_no, masked, label))
                                
    except Exception as e:
        print(f"[WARN] Could not read file {filepath}: {e}")
        
    return violations

def audit_patient_data_structures(filepath: str) -> List[str]:
    """
    Specifically validates patient data models (.json files) to ensure
    they follow HIPAA de-identification (no real names or addresses).
    """
    issues = []
    if not filepath.endswith(".json"):
        return issues
        
    # Known safe mock names and clinical categories that fit Name format
    allowed_names = {
        "Robert Davis", "Sarah Jenkins", "William Henderson", 
        "Global Sentinel", "Maternal Sentinel", "Pediatric Sentinel", "Geriatric Sentinel",
        "Systemic Health", "Oxidative Stress", "Antioxidant Status", "Alexander Vance", "Mara Santos",
        "Frida Kahlo", "Charles Darwin", "Srinivasa Ramanujan"
    }
        
    try:
        with open(filepath, "r", encoding="utf-8") as f:
            data = json.load(f)
            
        # Helper to inspect objects recursively
        def check_node(node):
            if isinstance(node, dict):
                # HIPAA Check: ensure clinical JSON data uses structured codes/mock IDs, not real PII keys
                for k, v in node.items():
                    if k.lower() in ("ssn", "socialsecurity", "birthdate") and v:
                        issues.append(f"Found sensitive key '{k}' in JSON structure.")
                    if isinstance(v, str):
                        # Simple name heuristic (Capitalized First + Last, e.g. 'John Doe')
                        if k.lower() == "name" and re.match(r"^[A-Z][a-z]+\s[A-Z][a-z]+$", v):
                            # Refined heuristic: patient names reside in dicts containing other profile keys
                            # (like 'age', 'gender', 'history', etc.) and not in sub-structures 
                            # (like 'dose', 'value', 'pathway', 'unit', etc.)
                            is_patient_name_prop = any(prop in node for prop in ("age", "gender", "history", "vitals", "scans"))
                            if is_patient_name_prop:
                                if not is_safe_placeholder(v) and v not in allowed_names:
                                    issues.append(f"Potential real patient name '{v}' in JSON property '{k}'")
                    check_node(v)
            elif isinstance(node, list):
                for item in node:
                    check_node(item)
                    
        check_node(data)
    except json.JSONDecodeError:
        pass # Not a valid json file, skip structure checks
    except Exception as e:
        print(f"[WARN] Could not parse JSON in {filepath}: {e}")
        
    return issues

def is_third_party_or_build(relative_path: str) -> bool:
    parts = relative_path.split(os.sep)
    vendor_dirs = {
        "node_modules", ".git", ".angular", "dist", "playwright-report",
        "test-results", "tmp", ".husky", "build", ".dart_tool", "ios", "android",
        "windows", "linux", "macos", "web", "sandbox", "flutter", "venv", ".venv", ".vscode"
    }
    if any(part in vendor_dirs for part in parts):
        return True
    if parts[-1] == "CMakeLists.txt" or "build_log" in parts[-1]:
        return True
    return False

def main():
    workspace_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    print(f"[START] Starting HIPAA/PII & Secret Security Scan in: {workspace_dir}\n")
    
    total_scanned = 0
    pii_violations_count = 0
    secret_violations_count = 0
    json_issues_count = 0
    
    for root, dirs, files in os.walk(workspace_dir):
        # Prune build/third party directories to speed up os.walk
        dirs[:] = [d for d in dirs if not is_third_party_or_build(os.path.relpath(os.path.join(root, d), workspace_dir))]
        
        for file in files:
            # Skip environment files and lock files
            if file.startswith(".env") or file in IGNORE_FILES:
                continue
            filepath = os.path.join(root, file)
            ext = os.path.splitext(file)[1].lower()
            
            if ext in SKIP_EXTENSIONS:
                continue
                
            relative_path = os.path.relpath(filepath, workspace_dir)
            total_scanned += 1
            
            if is_third_party_or_build(relative_path):
                continue
            
            # File content scanning
            violations = scan_file(filepath)
            if violations:
                for v_type, line_no, _, desc in violations:
                    print(f"[FAIL] [{v_type}] {relative_path}:{line_no} -> Found {desc} [match redacted]")
                    if v_type == "PII/PHI":
                        pii_violations_count += 1
                    else:
                        secret_violations_count += 1
                        
            # Custom JSON structure compliance checks
            json_issues = audit_patient_data_structures(filepath)
            if json_issues:
                for issue in json_issues:
                    print(f"[WARN] [HIPAA SCHEMA] {relative_path} -> {issue}")
                    json_issues_count += 1
                    
    print("\n" + "="*50)
    print("Scan Summary:")
    print(f"Scanned Files: {total_scanned}")
    print(f"  PII/PHI Violations: {pii_violations_count}")
    print(f"  Secrets/API Keys: {secret_violations_count}")
    print(f"  HIPAA Schema Issues: {json_issues_count}")
    print("="*50)
    
    if pii_violations_count > 0 or secret_violations_count > 0 or json_issues_count > 0:
        print("\n[FAIL] Compliance check failed! Please address the security findings above.")
        sys.exit(1)
    else:
        print("\n[PASS] Compliance scan passed. No PII/PHI or credential leaks detected.")
        sys.exit(0)

if __name__ == "__main__":
    main()
